src/coreclr/src/jit/morph.cpp

```diff
@@ -13705,24 +13716,56 @@ GenTree* Compiler::fgMorphSmpOp(GenTree* tree, MorphAddrContext* mac)
-                        else if (varTypeIsSmall(typ) && (genTypeSize(lvaTable[lclNum].lvType) == genTypeSize(typ)) &&
-                                 !lvaTable[lclNum].lvNormalizeOnLoad())
+                        bool matchingTypes  = (lclType == typ);
+                        bool matchingWidths = (genTypeSize(lclType) == genTypeSize(typ));
+                        bool matchingFloat  = (varTypeIsFloating(lclType) == varTypeIsFloating(typ));
+
+                        // TYP_BOOL and TYP_UBYTE are also matching types
+                        if (!matchingTypes)
                         {
-                            tree->gtType = typ = temp->TypeGet();
-                            foldAndReturnTemp  = true;
+                            if ((lclType == TYP_BOOL) && (typ == TYP_UBYTE))
+                            {
+                                matchingTypes = true;
+                            }
+                            else if ((lclType == TYP_UBYTE) && (typ == TYP_BOOL))
+                            {
+                                matchingTypes = true;
+                            }
                         }
-                        else if (!varTypeIsStruct(typ) && (lvaTable[lclNum].lvType == typ) &&
-                                 !lvaTable[lclNum].lvNormalizeOnLoad())
+
+                        // If the sizes match and we don't have a float or TYP_STRUCT,
+                        // then we will fold the IND/ADDR and use the LCL_VAR directly
+                        //
+                        if (matchingWidths && matchingFloat && (typ != TYP_STRUCT))
                         {
+                            // For some small types we might need to change to normalize loads or insert a cast here
+                            //
+                            if (varTypeIsSmall(typ))
+                            {
+                                // For any stores of small types, we will force loads to be normalized
+                                // this is necessary as we need to zero/sign extend any load
+                                // after this kind of store.
+                                //
+                                if (isStore)
+                                {
+                                    varDsc->lvForceLoadNormalize = true;
+                                }
+                                // otherwise we have a load operation
+                                //
+                                // Do we have a non matching small type load?
+                                // if so we need to insert a cast operation
+                                else if (!matchingTypes)
+                                {
+                                    addCastToType = typ;
+                                }
+                            }
+                            // we will fold the IND/ADDR and reduce to just the local variable
```

Comment:
just to make sure: is it intentional that with this PR we will fold `IND uint(ADDR(LCL_VAR int))` to `LCL_VAR uint` when before we would not do this? Is not it dangerous for RC1? Is it why we see improvements?

Reply:
Yes, it both prevents any regressions and improves codegen in a few cases.

Reply:
We did fold manyof these cases before, perhaps including the one you have written.

Reply:
I have the full framework diffs  above:

`164 total methods with Code Size differences (118 improved, 46 regressed), 227498 unchanged.`

Reply:
I think the previous check `else if (lvaTable[lclNum].lvType == typ))` would reject `UINT == INT` and such change makes me worried in the current release stage, is it possible to fix the bug without doing additional optimizations now and postpone the optimization to 6.0?

Reply:
I am confident that this change is an improvement and feel that it should go in as it is.

Reply:
You understand that  if (varTypeIsSmall(typ)) returns false for TYP_UINT right?

Reply:
I will spend some time investigating this chnage.

Reply:
I would tend to agree; it would probably be best to at least separate the optimization from the fix.